### PR TITLE
Add Demos for DSA params/DSA keygen.

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Fixes #19845

**Comment by slontis** [comment]
I have added these as demo code rather than bloating the pod files with documentation that will at some point be broken.
As you can load params via a PEM file i dont think this is needed in the pod.

File: demos/pkey/EVP_PKEY_DSA_keygen.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        goto cleanup;
+    }
+cleanup:
+    EVP_PKEY_CTX_free(ctx);
+    return dsaparamkey;
+}
+
+int main(int argc, char **argv)
+{
+    int rv = 1;
```
Comment (paulidale) [inline demos/pkey/EVP_PKEY_DSA_keygen.c:48]: `int rv = EXIT_FAILURE;`

File: demos/pkey/EVP_PKEY_DSA_keygen.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    if (EVP_PKEY_keygen_init(ctx) <= 0
+            || EVP_PKEY_keygen(ctx, &dsakey) <= 0) {
+        fprintf(stderr, "DSA keygen failed\n");
+        goto cleanup;
+    }
+
+    if (!dsa_print_key(dsakey, 1, libctx, propq))
+        goto cleanup;
+
+    rv = 0;
```
Comment (paulidale) [inline demos/pkey/EVP_PKEY_DSA_keygen.c:77]: `rv = EXIT_SUCCESS;`

Same for all the other programs.

File: demos/pkey/dsa.inc [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    0xbe, 0x14, 0x44, 0x1b, 0xd0, 0x93, 0xae, 0x66,
+    0xe5, 0x97, 0x6d, 0x53, 0x73, 0x3f, 0x4f, 0xa3,
+    0x26, 0x97, 0x01, 0xd3, 0x1d, 0x23, 0xd4, 0x67
+};
+
+/* Helper function to retrieve and print a key BIGNUM field */
+static void print_bn(BIO *bio, const EVP_PKEY *pkey, const char *name)
+{
+    BIGNUM *bn = NULL;
+    if (EVP_PKEY_get_bn_param(pkey, name, &bn) == 0)
```
Comment (paulidale) [inline demos/pkey/dsa.inc:93]: Blank line above

**Review by paulidale — APPROVED** [review]

File: demos/README.txt [hunk 4]
```diff
@@ -37,11 +37,14 @@ poly1305.c             Demonstration of Poly1305-AES message authentication
 siphash.c              Demonstration of SIPHASH message authentication
 
 pkey:
-EVP_PKEY_EC_keygen.c   Generate an EC key.
-EVP_PKEY_RSA_keygen.c  Generate an RSA key.
+EVP_PKEY_EC_keygen.c           Generate an EC key.
+EVP_PKEY_RSA_keygen.c          Generate an RSA key.
+EVP_PKEY_DSA_keygen.c          Generate an DSA key.
+EVP_PKEY_DSA_paramgen.c        Generate an DSA param key.
```
Comment (t8m) [inline demos/README.txt:43]: Nit: `a DSA....`

File: demos/README.txt [hunk 5]
```diff
@@ -37,11 +37,14 @@ poly1305.c             Demonstration of Poly1305-AES message authentication
...
+EVP_PKEY_DSA_paramvalidate.c   Validate a DSA param key.
+EVP_PKEY_DSA_paramfromdata.c   Load a DSA param key using raw data.
 
 smime:
-
```
Comment (t8m) [inline demos/README.txt:44]: Either not touch this part or provide the description for the demos in smime subdir.

File: demos/pkey/EVP_PKEY_DSA_paramvalidate.c [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * Example showing how to validate DSA parameters.
+ *
+ * Proper FIPS 186-4 DSA (FFC) parameter validation requires that all
+ * the parameters used for during parameter generation are supplied
```
Comment (t8m) [inline demos/pkey/EVP_PKEY_DSA_paramvalidate.c:14]: Nit: either `for` or `during`

**Review by paulidale — APPROVED** [review]

File: demos/README.txt [hunk 7]
```diff
@@ -37,14 +37,15 @@ poly1305.c             Demonstration of Poly1305-AES message authentication
 siphash.c              Demonstration of SIPHASH message authentication
 
 pkey:
-EVP_PKEY_EC_keygen.c           Generate an EC key.
-EVP_PKEY_RSA_keygen.c          Generate an RSA key.
-EVP_PKEY_DSA_keygen.c          Generate an DSA key.
-EVP_PKEY_DSA_paramgen.c        Generate an DSA param key.
+EVP_PKEY_EC_keygen.c           Generate a EC key.
+EVP_PKEY_RSA_keygen.c          Generate a RSA key.
```
Comment (t8m) [inline demos/README.txt:41]: Hmm.... I don't think these two should be changed to `a`.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
The fixup is trivial enough to assume Pauli's approval still holds.

**Comment by t8m** [comment]
Merged to master, 3.1, 3.0 branches. Thank you.

**Pull request closed without merging**